libs/networking/Testing_client.ml

```diff
@@ -0,0 +1,118 @@
+module Header = Cohttp.Header
+
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+
+type test_response = { response : Response.t; body_path : string }
+type make_response_fn = Request.t -> Body.t -> test_response Lwt.t
+
+module type S = sig
+  val make_response : make_response_fn
+end
+
+module Make (M : S) : Cohttp_lwt.S.Client = struct
+  open M
+  include Cohttp_lwt_unix.Client
+
+  let call ?(ctx = Net.default_ctx) ?headers ?(body = `Empty) ?chunked meth uri
+      =
+    ignore ctx;
+    let headers =
+      match headers with
+      | None -> Header.init ()
+      | Some h -> h
+    in
+    let chunked =
+      match chunked with
+      | Some c -> c
+      | None -> false
+    in
+    let req = Request.make_for_client ~headers ~chunked meth uri in
+    let%lwt response = make_response req body in
+    let response_body =
+      response.body_path |> Common.read_file |> Cohttp_lwt.Body.of_string
+    in
+    Lwt.return (response.response, response_body)
+
+  let get ?ctx ?headers uri = call ?ctx ?headers `GET uri
+
+  let delete ?ctx ?body ?chunked ?headers uri =
+    call ?ctx ?headers ?body ?chunked `DELETE uri
+
+  let post ?ctx ?body ?chunked ?headers uri =
+    call ?ctx ?headers ?body ?chunked `POST uri
+
+  let put ?ctx ?body ?chunked ?headers uri =
+    call ?ctx ?headers ?body ?chunked `PUT uri
+
+  let patch ?ctx ?body ?chunked ?headers uri =
+    call ?ctx ?headers ?body ?chunked `PATCH uri
+end
+
+(*****************************************************************************)
+(* Helper Functions *)
+(*****************************************************************************)
+
+let basic_response ?(status = `OK) ?(headers = Header.init ()) body_path =
+  let response = Response.make ~status ~headers ~flush:true () in
+  { response; body_path }
+
+let check_body body path =
+  let expected_body = path |> Common.read_file in
+  let%lwt actual_body = Cohttp_lwt.Body.to_string body in
+  Alcotest.(check string) "body" expected_body actual_body;
+  Lwt.return_unit
+
+let check_method req meth =
+  Alcotest.(check string)
+    "method" meth
+    (Cohttp.Request.meth req |> Cohttp.Code.string_of_method)
+
+let check_header req header header_val =
+  let actual_header = Cohttp.Header.get (Cohttp.Request.headers req) header in
+  match actual_header with
+  | None ->
+      Alcotest.fail
+        (Printf.sprintf "header %s not found. Headers: %s" header
+           (Cohttp.Header.to_string (Cohttp.Request.headers req)))
+  | Some actual_header ->
+      Alcotest.(check string) "header" header_val actual_header
+
+(*****************************************************************************)
+(* Entrypoint *)
+(*****************************************************************************)
+
+let with_testing_client make_fn test_fn () =
+  let prev_client = !Http_helpers.client_ref in
```

Comment:
You should use Common.save_excursion (it's a function I wrote inspired by the same function in emacs lisp).

Reply:
I didn't know about that, very cool